gnosis/verification.k

```diff
@@ -72,22 +72,52 @@ module VERIFICATION
 
 // Reading symbolic WordStack from memory, then extracting bytes from it. General case (stack size can be != 32)
 //========================================================================
-    rule nthbyteof(chop(#asWord(W : WS) &Int MASK), I       , N       )
-      => nthbyteof(chop(#asWord(    WS) &Int MASK), I -Int 1, N -Int 1)
-      requires I >Int 0
+    //decoding nthbyteof(#asWord())
+    rule nthbyteof(#asWord(W : WS), I       , N       )
+      => nthbyteof(#asWord(    WS), I -Int 1, N -Int 1)
+      requires I >Int 0 andBool #sizeWordStack(W : WS) <=Int 32
 
-    rule nthbyteof(chop(#asWord(W : WS) &Int MASK), 0, N) => W &Int ((MASK >>Byte #sizeWordStack(WS)) &Int 255)
+    rule nthbyteof(#asWord(W : WS), 0, N) => W
+      requires #sizeWordStack(W : WS) <=Int 32
+
+    //decoding nthbyteof(#asWord() &Int MASK)
+    rule nthbyteof(#asWord(W : WS) &Int MASK, I       , N       )
+      => nthbyteof(#asWord(    WS) &Int MASK, I -Int 1, N -Int 1)
+      requires I >Int 0 andBool #sizeWordStack(W : WS) <=Int 32
+
+    rule nthbyteof(#asWord(W : WS) &Int MASK, 0, N) => W &Int (MASK >>Byte #sizeWordStack(WS))
+      requires #sizeWordStack(W : WS) <=Int 32
+
+    //decoding nthbyteof() &Int MASK
+    rule MASK &Int nthbyteof(WS, I, N) => nthbyteof(WS, I, N) &Int MASK
+      requires #isConcrete(MASK)
+
+    rule nthbyteof(_, _, _)  &Int (A => A &Int 255)
+      requires A >Int 255
 
     rule nthbyteof(WS, I, N) &Int 255 => nthbyteof(WS, I, N)
 
-    rule nthbyteof(chop(MASK &Int #asWord(WS)) => chop(#asWord(WS) &Int MASK), _, _)
+    //decoding #asWord() &Int MASK
+    rule MASK &Int #asWord(WS) => #asWord(WS) &Int MASK
       requires #isConcrete(MASK)
       //version below doesn't work
       //requires notBool (getKLabel(MASK) ==K #klabel(`#asWord`))
 
+    //#asWord(...) & 255 is the last byte.
+    rule #asWord( X : Y : WS => Y : WS ) &Int 255
+    rule #asWord( X : .WordStack       ) &Int 255 => X
+
+    // #asWord only keeps 32 bytes and chops the rest.
+    rule #asWord(W : WS => WS)
+      requires #sizeWordStack(W : WS) >Int 32
 
-    syntax WordStack ::= #encSig (Int, Int, Int) [function]
+
+    syntax WordStack ::= #bytesSegment (Int, Int, Int, Int) [function] //DATA, START, RANGE_LEN, DATA_LEN
  // ===============================================================
-    rule #encSig(SIG_R, SIG_S, SIG_V) => #enc(#bytes32(SIG_R)) ++ #enc(#bytes32(SIG_S)) ++ #getValue(#uint8(SIG_V)) : .WordStack
+    rule #bytesSegment(DATA, START, LEN, N)
+         => nthbyteof(DATA, START, N) : #bytesSegment(DATA, START +Int 1, LEN -Int 1, N)
+      requires #range ( 0 <= START < N ) andBool #range( 1 <= START +Int LEN <= N ) andBool LEN >Int 0
```

Comment:
this could be double-edge sword.  sometimes we may want to keep the whole #bytesSegment term as it is, especially when we reason about the variable length buffer.  keep that in mind.